Warhead refactor - part 2: ValidImpactTypes

reaperrr opened the pull request:
This is basically https://github.com/OpenRA/OpenRA/pull/16312 done right, adressing its shortcomings and implementing some warhead generalizations discussed in that PR and on IRC.

#16312 can be removed from 'Future' milestone once this is merged.

Note1: I do plan to make `LeaveSmudgeWarhead` and `CreateResourceWarhead` `ValidImpactTypes`-aware in a follow-up, but I wanted to keep that out of this PR to keep possible distractions and delays to a minimum.
With this PR, those 2 warheads should just continue to work like on bleed, so there's no immediate pressure to adapt them here.

Note2: Turned the former testcase for RA into a fully-fledged refactor of the yaml setup, since there would have been a few regressions otherwise, afterall.

reaperrr committed:
Simplify CreateEffectWarhead

Simplified and streamlined code, moved everything
that might be usable by other warheads to Warhead.

reaperrr committed:
Refactor RA warhead setup

The old setup was not compatible
with the new, simplified warhead code.

Aircraft husks need a valid Targetable
now, because the explosions' parent actor
would otherwise be considered an invalid
actor target, blocking the terrain impact
but also return 'false' on the actor impact
warhead.

reaperrr commented:
I'm aware of the test errors, working on an update.

reaperrr committed:
Adapt TS effect warheads to ValidImpactTypes

reaperrr committed:
Remove ImpactActors boolean

teinarss commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -43,77 +43,14 @@ public class CreateEffectWarhead : Warhead
 		[Desc("Chance of impact sound to play.")]
```
Can remove the following usings from this file:
using System.Linq;
using OpenRA.Mods.Common.Traits;
using OpenRA.Primitives;

teinarss commented:
#16727 happens for me on this PR.

reaperrr commented:
Updated (and rebased).

> #16727 happens for me on this PR.

Should be fixed now.
At least in RA, the 'testcase' setup was necessary for all weapons with actor-/terrain-dependant effect warheads, afterall. The "submerged sub" edge case prevents any easier solution, and I didn't want to re-introduce the `ImpactActors` switch on the code side.
Merged the DepthCharge fix and testcase into that RA effect WH refactor, so this PR is now mergeable as-is (once any other bugs/style issues are found and fixed).

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Warheads/Warhead.cs:
```diff
@@ -42,6 +38,18 @@ public abstract class Warhead : IWarhead
+			"Current options are Terrain and Actor. If only Terrain is listed, warhead will never trigger on actors.",
+			"If only Actor is listed, the warhead will never trigger on empty terrain.",
+			"If both are listed, the warhead will trigger if either any touched actor or the underlying terrain has a valid target type.")]
+		public readonly ImpactType ValidImpactTypes = ImpactType.Terrain | ImpactType.Actor;
```
Would it make sense to add an `ImpactType.Parent` to replace `AffectsParent` below? Or do you think that would be better solved by overhauling `ValidStances` instead? Or neither?

GraionDilach replied:
Neither. Parent has nothing to do with stances (usecase: chain-bolt healing effect targeting, if it allows selftargeting back, the owner is encouraged to micro out 2 unis to bounce the healing back to selfheal, if self is prevented, it is basically just a fancy, "smart" area-of-effect) and would be inconsistent with Actor (since Parent is an Actor).

pchote commented:
Pulling together my musings from IRC into something useful, my main concern is that the current impact types are tailored to our legacy projectile/warhead model that forces everything to be applied via spread damage, and that they aren't expressive enough to work properly once we break away from that.

We could solve this by defining a more general set of `ImpactType`s, and specify both `ValidImpactTypes` and `InvalidImpactTypes`. The `IsValidImpact` function would take a `Target` instead of a position, and first check for any matches in `InvalidImpactTypes` and return false. It would then check for any matches in `ValidImpactTypes` and return true. If nothing matches it returns false.

The expanded impact types could be:
   * `ValidTerrain`: (`Target.Type == TargetType.Terrain` and terrain matches `ValidTargets`)
   * `InvalidTerrain`: (`Target.Type == TargetType.Terrain` and terrain matches `InvalidTargets`)
   * `ValidActor`: (`Target.Type == TargetType.Actor` and target matches `ValidTargets`)
   * `InvalidActor`: (`Target.Type == TargetType.Actor` and target matches `InvalidTargets`)
   * `TerrainWithValidActor` (`Target.Type == TargetType.Terrain` and `FindActorsOnCircle` returns an actor that matches `ValidTargets`)
   * `TerrainWithInvalidActor` (`Target.Type == TargetType.Terrain` and `FindActorsOnCircle` returns an actor that matches `InvalidTargets`)
  * `Air`: (detonation altitude > `AirThreshold`, no nearby actors)
  * `AirWithValidActor`: (detonation altitude > `AirThreshold`, nearby actor matching `ValidTargets`)
  * `AirWithInvalidActor`: (detonation altitude > `AirThreshold`, nearby actor matching `InvalidTargets`)

and `TargetTypeAir` field would go way. `TerrainWithValidActor` in `ValidTargets` would take priority over `InvalidTerrain` in `InvalidTargets`, and likewise `TerrainWithInvalidActor` in `InvalidTargets` would take priority over `Terrain` in `ValidTargets`.

The defaults would be:
```
ValidImpacts = ImpactType.ValidTerrain | ImpactType.ValidActor | ImpactType.TerrainWithValidActor
InvalidImpacts = ImpactType.InvalidTerrain | ImpactType.InvalidActor
```
which should hopefully be equivalent to what we have now (remember that the new `TerrainWithValidActor` is equivalent to the current `Actor`, and the new `Actor` only applies to projectiles like `InstantHit`.


The water/ship/submarine cases would be covered by:
```
Warhead@EffDirectHit: CreateEffect
		Explosions: small_explosion
		ImpactSounds: kaboom12.aud
		ValidImpactTypes: ValidTerrain, TerrainWithValidActor, Air
		ValidImpactTypes: InvalidTerrain, TerrainWithInvalidActor
		AffectsParent: true
		ValidTargets: Ground, Air, Trees
		InvalidTargets: Water, Underwater
	Warhead@EffWater: CreateEffect
		Explosions: small_splash
		ImpactSounds: splash9.aud
		ValidImpactTypes: ValidTerrain
		ValidImpactTypes: InvalidTerrain, TerrainWithInvalidActor
		ValidTargets: Water, Underwater
		InvalidTargets: Ground
```

It would probably be helpful to consider more cases before trying to implement any changes.

pchote commented:
Or a different approach could be to come up with a way to have the warheads trigger the effects directly (perhaps implementing a new concept for visual effects that isn't a projectile or a warhead?) - this would let us contain the uglyness of the spread damage logic within the `SpreadDamageWarhead` so that it doesn't affect the more fundamental warhead concepts.

reaperrr commented:
> It would probably be helpful to consider more cases before trying to implement any changes.

I think RA with its ships/subs and TS with hover units over water covers the most problematic cases, so in my opinion a test branch would likely give the best impression of whether this works as intended.

> Or a different approach could be to come up with a way to have the warheads trigger the effects directly (perhaps implementing a new concept for visual effects that isn't a projectile or a warhead?)

That may be a good idea independently, but I'm not sure how that would help with regards to the complexity of this PR (or the impact type checks in general)?

> once we break away from that.

Could you give me a rough idea how that might look like?
Or how soon you think that should/might happen?
Because...

> this would let us contain the uglyness of the spread damage logic within the SpreadDamageWarhead so that it doesn't affect the more fundamental warhead concepts.

This PR and specifically moving the impact type stuff from CEWH to base Warhead was in part meant as preparation for some changes needed for the Cyborg Reaper and ClusterMissile-related warheads and improvements (though it's been so long I don't remember the specifics) and possibly some more downstream stuff, so it would be good to figure out alternatives and implement them soon if you really feel the current approach isn't good enough even as mid-term solution.

But my concern here is that at the 'pace' we're sometimes moving in terms of reviews and getting stuff merged, it may well be 2 or 3 years or more away until that actually happens, and will likely require a larger refactor and update rule anyway, so in my opinion it would be the lesser evil to have a bit more work in the future, but a good-enough (for modders) improvement until that happens.

In any case, I'll try the "expanded impact types" approach first. But I now doubt this will move along quickly, so I'll decouple #17119 and make it not depend on this.

pchote commented:
> Could you give me a rough idea how that might look like?
> Or how soon you think that should/might happen?

Bad wording on my part: I meant this in a theoretical / other mods sense rather than any imminent plans for the default mods.

> with regards to the complexity of this PR (or the impact type checks in general)?

The issue isn't really with the complexity, more with expressiveness/elegance. We've been able to find representations for many of our other systems (e.g. traits, sequences, effects, projectiles) that are fairly intuitive and mostly avoid imposing specific implementation details. Its been a bugbear of mine that we haven't been able to achieve this for warheads, and that every new feature seems to require major structural changes to the way they work. #17119 is a step in the right direction.

Mailaender commented on mods/ra/weapons/missiles.yaml:
```diff
@@ -338,8 +362,13 @@ SCUD:
+		ImpactSounds: firebl3.aud
+	Warhead@EffSubmergedHit: CreateEffect
+		Explosions: large_splash
+	Warhead@EffWater: CreateEffect
```
Why does this have no
```
ValidTargets: 
InvalidTargets:
```
?

reaperrr replied:
`^SubMissileDefault` inherits them from `^AntiGroundMissile`.

reaperrr commented:
I'm planning to get back to this eventually, but it would likely need an overhaul and shouldn't block Next+1. Closing for now and moving to Future milestone.
